src/common/bitstream.h

```diff
@@ -198,6 +201,19 @@ class BitStreamImpl : boost::noncopyable, public BitStream {
 		return v;
 	}
 
+	/** Read a multi-bit signed value from the bit stream. */
+	int32 getSignedBits(size_t n) {
+		if (n == 0)
+			return 0;
+
+		int32 v = getBits(n);
+
+		if (v >> (n - 1))
+			v |= 0xFFFFFFFF << n;
```

Comment:
Technically, that might produce a value that traps on some negative value representations.

What is it you want to accomplish? Isn't it just a masking away of the sign bit and negating the value? 

Why don't you just do a & with the ~ of the bit shifted to the correct position (on the unsigned value), cast to signed and then unary-minus the whole thing? That makes it way clearer what's going on, and is more portable.

Reply:
I mean, all these bit fiddling tricks you'd normally do if you want to get rid of the branch, but you've got an explit branch with the if there anyway...